[k8s][docs] K3s docs and runtime auto-detection

romilbhardwaj opened the pull request:
Closes #3034 and #2958. Adds docs and automatic runtimeclass detection for k3s users. 

Tested (run the relevant ones):

- [x] Docs rendered locally
- [x] Code formatting: `bash format.sh`
- [x] Tested GPU labelling script and `sky launch` on a manually setup k3s cluster
- [x] Tested `sky launch` on GKE cluster
- [x] Kubernetes smoke tests using T4.

Michaelvll commented on sky/utils/kubernetes/gpu_labeler.py:
```diff
@@ -105,6 +119,22 @@ def label():
+        except Exception as e:  # pylint: disable=broad-except
+            print(f'Error occurred while checking for nvidia RuntimeClass: {e}')
+            print('Continuing without using nvidia RuntimeClass. '
+                  'This may fail on K3s clusters.')
```
```suggestion
                  'This may fail on K3s clusters. For more details, refer to <link-to-our-doc>')
```

Michaelvll commented on docs/source/reference/kubernetes/kubernetes-setup.rst:
```diff
@@ -181,9 +181,17 @@ If your Kubernetes cluster has Nvidia GPUs, ensure that:
-    If you are using RKE2, the GPU operator installation through helm requires extra flags to set ``nvidia`` as the default runtime for containerd. Refer to instructions on `Nvidia GPU Operator installation with Helm on RKE2 <https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/getting-started.html#custom-configuration-for-runtime-containerd>`_ for details.
+
+    GPU labels are case-sensitive. Ensure that the GPU name is lowercase if you are using the ``skypilot.co/accelerators`` label.
+
```
.. _automatic-gpu-labelling:

Michaelvll commented on docs/source/reference/kubernetes/kubernetes-setup.rst:
```diff
@@ -197,23 +205,93 @@ We provide a convenience script that automatically detects GPU types and labels
+    handler: nvidia
+    EOF
+
+Now you can label your K3s nodes with the :code:`skypilot.co/accelerator` label using the SkyPilot GPU labelling script above.
```
```suggestion
Now you can label your K3s nodes with the :code:`skypilot.co/accelerator` label using the `SkyPilot GPU labelling script <automatic-gpu-labelling>`_ above.
```

Michaelvll reviewed (commented):
This is awesome @romilbhardwaj! Left some nits.

Michaelvll commented on sky/utils/kubernetes/gpu_labeler.py:
```diff
@@ -30,6 +30,20 @@ def prerequisite_check() -> Tuple[bool, str]:
     return prereq_ok, reason
 
 
+def check_nvidia_runtime_class() -> bool:
+    """Checks if the 'nvidia' RuntimeClass exists in the cluster"""
+    config.load_kube_config()
+    node_api = client.NodeV1Api()
+
+    # Fetch the list of available RuntimeClasses
+    runtime_classes = node_api.list_runtime_class()
+
+    # Check if 'nvidia' RuntimeClass exists
+    nvidia_exists = any(
+        rc.metadata.name == 'nvidia' for rc in runtime_classes.items)
+    return nvidia_exists
```
Should we just use the functino in `sky.provision.kubernetes.instance` instead?

romilbhardwaj replied:
good idea- done!

Michaelvll reviewed (approved):
Thanks for the quick fix @romilbhardwaj! LGTM.

romilbhardwaj commented:
Thanks @Michaelvll! I've refactored and tested the gpu_labeller on k3s and EKS. 

I've also added a `deploy_k3s.sh` for our debugging purposes to easily deploy a k3s cluster for testing.
